Buildings/Fluid/Storage/Plant/Controls/TankStatus.mo

```diff
@@ -0,0 +1,195 @@
+  parameter Modelica.Units.SI.TemperatureDifference dTUnc = 0.1
+    "Temperature sensor uncertainty";
+  parameter Modelica.Units.SI.TemperatureDifference dTHys = 1
+    "Deadband for hysteresis";
+
+  Modelica.Blocks.Interfaces.RealInput TTan[2](
+    each final quantity="Temperature",
+    each displayUnit="C")
+    "Temperatures at the tank 1: top and 2: bottom" annotation (Placement(
+        transformation(
+        extent={{-10,-10},{10,10}},
+        rotation=0,
+        origin={-110,0}), iconTransformation(
+        extent={{-10,-10},{10,10}},
+        rotation=0,
+        origin={-110,0})));
+
+  inner Modelica.StateGraph.StateGraphRoot stateGraphRoot
+    annotation (Placement(transformation(extent={{-60,0},{-40,20}})));
+  Modelica.StateGraph.InitialStep iniSte(nOut=1, nIn=1)
+    "Initial step, also active when none of the other steps apply"
+    annotation (Placement(transformation(extent={{-60,-40},{-40,-20}})));
+  Modelica.StateGraph.Transition traDep(condition=TTan[2] > THig - dTUnc)
+    "Transition: Tank is depleted"
+    annotation (Placement(transformation(extent={{40,20},{60,40}})));
+  Modelica.StateGraph.Transition traCoo(condition=TTan[1] < TLow + dTHys)
+    "Transition: Tank is cooled"
+    annotation (Placement(transformation(extent={{40,-60},{60,-40}})));
+  Modelica.StateGraph.StepWithSignal steDep(nIn=1, nOut=1)
+    "Step: Tank is depleted"
+    annotation (Placement(transformation(extent={{80,20},{100,40}})));
+  Modelica.StateGraph.Alternative altDepOrCoo(nBranches=2)
+    "Alternative: Tank is depleted or cooled"
+    annotation (Placement(transformation(extent={{-18,-120},{338,60}})));
+  Modelica.StateGraph.StepWithSignal steCoo(nIn=1, nOut=1)
+    "Step: Tank is cooled"
+    annotation (Placement(transformation(extent={{80,-60},{100,-40}})));
+  Modelica.StateGraph.Alternative altOveCoo(nBranches=2)
+    "Alternative: Reset or tank is overcooled"
+    annotation (Placement(transformation(extent={{112,-100},{276,0}})));
+  Modelica.StateGraph.Transition traRes1(condition=TTan[2] < THig - dTHys)
+    "Transition: Reset to initial step"
+    annotation (Placement(transformation(extent={{180,20},{200,40}})));
+  Modelica.StateGraph.Transition traCoo2(condition=TTan[1] < TLow + dTUnc)
+    "Transition: Tank is cooled"
+    annotation (Placement(transformation(extent={{140,-80},{160,-60}})));
+  Modelica.StateGraph.Transition traRes2(condition=TTan[1] > TLow + dTHys*2)
+    "Transition: Reset to initial step"
+    annotation (Placement(transformation(extent={{180,-40},{200,-20}})));
+  Modelica.StateGraph.StepWithSignal steOveCoo(nIn=1, nOut=1)
+    "Step: Tank is overcooled (and is still cooled)"
+    annotation (Placement(transformation(extent={{180,-80},{200,-60}})));
+  Modelica.StateGraph.Transition traRes3(condition=traRes2.condition)
+    "Transition: Reset to initial step"
+    annotation (Placement(transformation(extent={{220,-80},{240,-60}})));
+  Modelica.Blocks.Interfaces.BooleanOutput y[3]
```

Comment:
I'm not sure why this is a vector of three booleans instead of an enumeration, which I think is more intuitive, unless there's a specific reason more than one needs to be true at one time?

Reply:
There is indeed a reason behind this. As the output of the `TankStatus` block, the different states of the tank are not clean cut between "cooled" and "overcooled". When the tank is "cooled" it may or may not be "overcooled", but whenever it is "overcooled" it must also be "cooled". Separating these signals make it easier and clearer for the `FlowControl` block to handle the overlapping states accordingly.
Using an enumeration/integer signal makes this less clear. For example, with an implementation that looks like
`1 = is depleted, 2 = intermediate, 3 = cooled (but not overcooled), 4 = overcooled`,
it becomes inevitable to separate `3` and `4`. Down the line in `FlowControl`, simple statements like
`if isCooled` and `if not isCooled` would become `if y == 3 or y == 4` and `if y <>3 and y <> 4`.